### PR TITLE
[MM-23017] Add check for groups to SendNotifications 

### PR DESCRIPTION
**Pull request by fmunshi**
#### Summary
- Adds logic to check for groups in the `SendNotifications` function (If user can "group mention")
- getExplicitMentions now expects a map of groups to be passed along side the existing map of user keywords
- Inserts users mentioned by any group mentions into the existing map of mentioned users

#### Notes:
- ~Since the `group_mentions` permission is not yet merged I am just using `PERMISSION_USE_GROUP_MENTIONS := model.PERMISSION_USE_CHANNEL_MENTIONS`~
- `PERMISSION_USE_GROUP_MENTIONS` is being checked now since it is merged in master

- ~`FilterAllowReference` is added in https://github.com/mattermost/mattermost-server/pull/14010 so I have some the `opts.FilterAllowReference` check commented out in `group_store`, I will eventually merge 14010 with this branch either here or in a separate feature branch~
- Checks for FilterAllowReference have been added 

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-23017

### CODE REVIEW

File: app/notification.go [hunk 1]
```diff
@@ -36,6 +36,22 @@ func (a *App) SendNotifications(post *model.Post, team *model.Team, channel *mod
 		close(cmnchan)
 	}()
 
+	var gchan chan store.StoreResult
+	if a.allowGroupMentions(post) {
```
Comment (fm2munsh) [inline app/notification.go]: If the user does not have access to group mentions no extra query will be made for groups and the groups map will just be empty

File: store/sqlstore/group_store.go [hunk 2]
```diff
@@ -312,6 +312,40 @@ func (s *SqlGroupStore) GetMemberCount(groupID string) (int64, *model.AppError)
 	return count, nil
 }
 
+func (s *SqlGroupStore) GetMemberUsersNotInChannel(groupID string, channelID string) ([]*model.User, *model.AppError) {
```
Comment (fm2munsh) [inline store/sqlstore/group_store.go]: Is this query too specific and should maybe be moved into `GetMemberUsers` with some options struct? to achieve the same behaviour?

Comment (hahmadia) [reply store/sqlstore/group_store.go]: I don't have any opinion on wether or not it should be moved but I would say it makes more sense if it was called `GetGroupMemberUseresNotInChannel`.

Comment (fm2munsh) [reply store/sqlstore/group_store.go]: The reason I called it `GetMemberUsersNotInChannel ` is to match the style of the `GetMemberUsers` function which gets all of a groups users

File: app/notification.go [hunk 3]
```diff
@@ -36,6 +36,22 @@ func (a *App) SendNotifications(post *model.Post, team *model.Team, channel *mod
...
+		gchan = make(chan store.StoreResult, 1)
+		go func() {
+			groups, err := a.Srv().Store.Group().GetGroups(0, 0, model.GroupSearchOpts{FilterAllowReference: true})
+			groupsMap := make(map[string]*model.Group)
+			if err == nil {
+				for _, group := range groups {
+					groupsMap[group.DisplayName] = group
```
Comment (catalintomai) [inline app/notification.go]: @fm2munsh , use group.Name instead?

Comment (fm2munsh) [reply app/notification.go]: yea - done

File: store/sqlstore/group_store.go [hunk 4]
```diff
@@ -1019,6 +1058,11 @@ func (s *SqlGroupStore) GetGroups(page, perPage int, opts model.GroupSearchOpts)
 		`, opts.NotAssociatedToChannel)
 	}
 
+	if opts.FilterAllowReference {
+		// Todo uncomment after column has been added
+		// 	groupsQuery = groupsQuery.Where("g.AllowReference = 1")
```
Comment (catalintomai) [inline store/sqlstore/group_store.go]: need to be "g.AllowReference = true"

Comment (fm2munsh) [reply store/sqlstore/group_store.go]: I made that change in the feature branch

File: app/notification.go [hunk 5]
```diff
@@ -639,6 +721,20 @@ func (a *App) allowChannelMentions(post *model.Post, numProfiles int) bool {
 	return true
 }
 
+// allowGroupMentions returns whether or not the group mentions are allowed for the given post.
+func (a *App) allowGroupMentions(post *model.Post) bool {
+	PERMISSION_USE_GROUP_MENTIONS := model.PERMISSION_USE_CHANNEL_MENTIONS
+	if !a.HasPermissionToChannel(post.UserId, post.ChannelId, PERMISSION_USE_GROUP_MENTIONS) {
```
Comment (fm2munsh) [inline app/notification.go]: This permission is changed to use the actual `PERMISSION_USE_GROUP_MENTIONS` in the feature branch - I have not done so here because it would cause the build to fail

File: app/notification_test.go [hunk 6]
```diff
@@ -911,6 +960,38 @@ func TestAllowChannelMentions(t *testing.T) {
...
+
+	t.Run("should return false for a channel purpose post", func(t *testing.T) {
+		purposeChangePost := &model.Post{ChannelId: th.BasicChannel.Id, UserId: th.BasicUser.Id, Type: model.POST_PURPOSE_CHANGE}
+		allowGroupMentions := th.App.allowGroupMentions(purposeChangePost)
+		assert.False(t, allowGroupMentions)
+	})
+
+	t.Run("should return false for a post where the post user does not have USE_GROUP_MENTIONS permission", func(t *testing.T) {
+		PERMISSION_USE_GROUP_MENTIONS := model.PERMISSION_USE_CHANNEL_MENTIONS
+		defer th.AddPermissionToRole(PERMISSION_USE_GROUP_MENTIONS.Id, model.CHANNEL_USER_ROLE_ID)
```
Comment (fm2munsh) [inline app/notification_test.go]: This permission is changed to use the actual `PERMISSION_USE_GROUP_MENTIONS` in the feature branch - I have not done so here because it would cause the build to fail

File: store/sqlstore/group_store.go [hunk 7]
```diff
@@ -312,6 +312,41 @@ func (s *SqlGroupStore) GetMemberCount(groupID string) (int64, *model.AppError)
...
+
+	query := `
+		SELECT
+			Users.*
+		FROM
+			GroupMembers
+			JOIN Users ON Users.Id = GroupMembers.UserId
+		WHERE
+			GroupMembers.DeleteAt = 0
+			AND Users.DeleteAt = 0
```
Comment (catalintomai) [inline store/sqlstore/group_store.go]: nit: I'd move this lines to the bottom to keep consistent to having DeleteAt checks at the end.

File: app/notification.go [hunk 8]
```diff
@@ -657,6 +753,44 @@ func (a *App) getMentionKeywordsInChannel(profiles map[string]*model.User, allow
 	return keywords
 }
 
+// insertGroupMentions adds Mentions for users that are part of the given group members and channel channel, adds group members that are not in the channel to OtherPotentialMentions
```
Comment (catalintomai) [inline app/notification.go]: maybe rephrase to say "for users that are members of the given group", etc.

File: app/notification.go [hunk 9]
```diff
@@ -657,6 +753,44 @@ func (a *App) getMentionKeywordsInChannel(profiles map[string]*model.User, allow
...
+// returns true if atleast one user from the group exists in the team that the channel belongs to.
```
Comment (catalintomai) [inline app/notification.go]: "exists in the team" - is it the team or the channel? I see that profileMap is built off GetAllProfilesInChannel() result.

Comment (fm2munsh) [reply app/notification.go]: yea its exists in team cause it looks at users in the channel from `profileMap` and then it gets the list of `outOfChannelGroupMembers` so in the case where there are no `outOfChannelGroupMembers` and no users mentioned from the `profileMap` it returns `false` to let us know that the group mention did nothing

File: app/notification.go (hunk 8 above)
Comment (fm2munsh) [reply app/notification.go]: the wording on this was horrible - ill fix it

File: app/notification.go (hunk 9 above)
Comment (catalintomai) [reply app/notification.go]: I was referring to "return true" in the comment , not "return false". Unless I miss something, `userMentioned` is set to true if there is at least one user from the channel, irrespective of the other conditions.

Comment (fm2munsh) [reply app/notification.go]:
```
for _, user := range users {
		if _, ok := profileMap[user.Id]; ok {
			mentions.Mentions[user.Id] = GroupMention
			userMentioned = true
		}
	}
```

`users` refers to group members here - i think the variable should be renamed to `groupMembers`

So it iterates through all the group member users and then for each one checks if the `profileMap` matches a user id if theres atleast one user found it sets `userMentioned` to true

File: app/notification.go [hunk 10]
```diff
@@ -657,6 +753,44 @@ func (a *App) getMentionKeywordsInChannel(profiles map[string]*model.User, allow
...
+// returns false if no group members present in the team that the channel belongs to
+func (a *App) insertGroupMentions(group *model.Group, channel *model.Channel, profileMap map[string]*model.User, mentions *ExplicitMentions) (bool, *model.AppError) {
+	userMentioned := false
+	groupMembers, err := a.Srv().Store.Group().GetMemberUsers(group.Id)
+	if err != nil {
+		return userMentioned, err
+	}
+
+	if mentions.Mentions == nil {
+		mentions.Mentions = make(map[string]MentionType)
```
Comment (fm2munsh) [inline app/notification.go]: As discussed with @catalintomai this will be replaced with a new store function for `GetMemberUsersInTeam` saving us from having to do the call to `.GetMemberUsersNotInChannel` however I will still commit that store function since it will be useful when we need to do group invites to channel

**Comment by fmunshi** [comment]
@srkgupta This PR is ready to be QA'd now, it deals exclusively with sending group members notifications for group mentions and no UI changes outside of the new system messages returned.

Some things to test:
- Groups mentioned with the permission to mention groups should trigger notifications
- Groups mentioned without the permission to mention groups should not trigger notifications
- Users in groups with `FilterAllowReference = true` should receive notifications when group is mentioned by name
- Users in groups with `FilterAllowReference = false` should not receive notifications when group is mentioned by name
- Group mentions should not trigger mentions when used between code formatting markdown 
- Group mentions when group has no members on the team should return a system message
- Group mentions when group has members on team but outside channel should return a system message asking to add users to channel

**Comment by fmunshi** [comment]
@srkgupta I am unable to reproduce 
"MM-24221 - User is still allowed to use Group Mentions and notify other users in the group even when the permission was removed from System Scheme"

Since I have made it to enable group mentions by default for team and channel admins you have to go in and manually remove for channel / team admins, can you confirm that the user you tested with was not a channel or team admin with the permission still enabled

**Comment by fmunshi** [comment]
@srkgupta I have resolved MM-24220

**Comment by srkgupta** [comment]
> @srkgupta I am unable to reproduce 
> "MM-24221 - User is still allowed to use Group Mentions and notify other users in the group even when the permission was removed from System Scheme"
> 
> Since I have made it to enable group mentions by default for team and channel admins you have to go in and manually remove for channel / team admins, can you confirm that the user you tested with was not a channel or team admin with the permission still enabled

Yes @Farhan I endured that the user is a regular member and not a channel or team admin
